scipy/_lib/_disjoint_set.py

```diff
@@ -0,0 +1,174 @@
+"""
+Disjoint set data structure
+"""
+import collections
+
+
+class DisjointSet:
+    """ Disjoint set data structure for incremental connectivity queries.
+
+    .. versionadded: 1.6.0
+
+    Attributes
+    ----------
+    n_subsets : int
+        The number of subsets.
+
+    Methods
+    -------
+    add
+    merge
+    connected
+    __getitem__
+
+    Notes
+    -----
+    This class implements the disjoint set [1]_, also known as the *union-find*
+    or *merge-find* data structure. The *find* operation (implemented in
+    `__getitem__`)* implements the *path halving* variant. The *merge* method
+    implements the *merge by size* variant.
+
+    References
+    ----------
+    .. [1] https://en.wikipedia.org/wiki/Disjoint-set_data_structure
+
+    See Also
+    --------
+    connected_components : Analyze the connected components of a sparse graph.
+
+    Examples
+    --------
+    >>> from scipy.sparse.csgraph import DisjointSet
+
+    Initialize a disjoint set:
+
+    >>> disjoint_set = DisjointSet([1, 2, 3, 'a', 'b'])
+
+    Merge some subsets:
+
+    >>> disjoint_set.merge(1, 2)
+    True
+    >>> disjoint_set.merge(3, 'a')
+    True
+    >>> disjoint_set.merge('a', 'b')
+    True
+    >>> disjoint_set.merge('b', 'b')
+    False
+
+    Find root elements:
+
+    >>> disjoint_set[2]
+    1
+    >>> disjoint_set['b']
+    3
+
+    Test connectivity:
+
+    >>> disjoint_set.connected(1, 2)
+    True
+    >>> disjoint_set.connected(1, 'b')
+    False
+
+    List elements in disjoint set:
+
+    >>> list(disjoint_set)
+    [1, 2, 3, 'a', 'b']
+    """
+    def __init__(self, elements=None):
+        self.n_subsets = 0
+        self._sizes = {}
+        self._parents = {}
+        self._indices = collections.OrderedDict()
+        if elements is not None:
+            for x in elements:
+                self.add(x)
+
+    def __iter__(self):
+        """Returns an iterator of the elements in the disjoint set.
+
+        Elements are ordered by insertion order.
+        """
+        return iter(self._indices)
+
+    def __len__(self):
+        return len(self._indices)
+
+    def __contains__(self, x):
+        return x in self._indices
+
+    def __getitem__(self, x):
+        """Find the root element of `x`.
+
+        Parameters
+        ----------
+        x : hashable object
+            Input element.
+
+        Returns
+        -------
+        root : hashable object
+            Root element of `x`.
+        """
+        if x not in self._indices:
+            raise KeyError(x)
+
+        # find by "path halving"
+        parents = self._parents
+        while self._indices[x] != self._indices[parents[x]]:
```

Comment:
Is this ever true with `x != parents[x]`?
Would `hash(x) != hash(parents[x])` be faster?

Reply:
`x != parents[x]` fails if `x` is e.g. a numpy integer. fuglede and I discussed this at some length above.

`hash` is susceptible to collisions:
```
>>> hash((1,0,0))
2528505496374819208
>>> hash((1,-2,-2))
2528505496374819208
```
(from a [lively debate](https://bugs.python.org/issue34751))

Reply:
Thanks for linking. I have more respect for Raymond, Tim, and Python core after reading about that journey.

Seem fudlede and you hashed this out already, I would not have thought of np.int / ndarray so great catch on `__eq__` being wonky.

`hash(x) == hash(parents[x]) and x == parents[x]` might be better but it isn't enought to overcome :/

Reply:
The `hash and __eq__` approach should be equivalent to using `_indices`, cf. the discussion in the other comment: If the hashes collide, an equality comparison is forced already in `not in`.